### PR TITLE
Rework outgoing message state.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen** [comment]
See: https://github.com/WhisperSystems/Signal-iOS/pull/1974

File: src/Messages/Interactions/TSOutgoingMessage.h
Comment (charlesmchen) [inline src/Messages/Interactions/TSOutgoingMessage.h:85]: These properties are used from multiple threads so they should all be atomic.

File: src/Messages/Interactions/TSOutgoingMessage.h
Comment (charlesmchen) [inline src/Messages/Interactions/TSOutgoingMessage.h]: This is the key change.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:494]: This is a significant change. When re-sending group messages that failed to safety number changes, we used a special code path to avoid double-sending to other members of the group.  That code path was mostly redundant with our normal message sending code path but had some differences. Most significantly it ignored the message send operation queue.  I've consolidated along the one true code path, for consistency and correctness.  We now avoid double-sending to other members of a group using another mechanism which also applies to message send retries.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:683]: We now do some bookkeeping to track which group members have received a given message.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:710]: ...And we use that bookkeeping to reduce (but not completely avoid) double-sending.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:941]: In https://github.com/WhisperSystems/SignalServiceKit/pull/167 we took a big step towards only marking messages as failed or succeeded in a single place and only when that state transition was "final".  It turns out there was more work to be done.  We now use the OWSSendMessageOperation's success and failure handlers to mark success and failure.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:1237]: I need to investigate more before I remove this change.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:1238]: I believe this method was redundant. It did two things:

* Save a new error message, which was never initialized, so this had no effect.
* Mark a message send as failed.  We want to do this in `OWSSendMessageOperation`'s failure handler, not here.

**Comment by charlesmchen** [comment]
I guess it's worth stating that these changes are sensitive and need to be extensively tested.

File: src/Messages/OWSMessageSender.m
Comment (michaelkirk) [reply src/Messages/OWSMessageSender.m:494]: I'll be really happy to get rid of this. =)

When this code was written, Android nor iOS had timestamp deduping in place, so it was a real problem.

Signal-Desktop is still missing timestamp deduping support, otherwise, I'd just say let's keep our code simple and blast it out to everyone.

So I think this approach is great for now. Maybe in the fully deduped future we can come back and simplify it.

**Comment by charlesmchen** [comment]
@michaelkirk I've uncovered more work to be done here. Please hold off on reviewing this PR.

**Comment by charlesmchen** [comment]
@michaelkirk I've uncovered more work to be done here. Please hold off on reviewing this PR.

**Comment by charlesmchen** [comment]
@michaelkirk This PR is now ready for review.

File: src/Messages/Interactions/TSOutgoingMessage.m
Comment (michaelkirk) [inline src/Messages/Interactions/TSOutgoingMessage.m]: The reason for this method may not be obvious outside of the context of this PR. Could you add a method comment to explain why it exists?

File: src/Messages/OWSMessageSender.m
Comment (michaelkirk) [inline src/Messages/OWSMessageSender.m]: As I understand this change, we're now going to mark a group send as failed, if a single recipient fails to receive it. 

That has an obvious up side. The downside is that we'll confuse the situation where multiple recipients do get it but, e.g. the one unregistered group member, does not get it.

Is there a way to address this without going down the design-rabbit-hole of per-recipient delivery status? Just brainstorming... like wrapping the error and displaying it in the retry copy or footer status.

Even if not, I think what you have is a more conservative (better in this case) approach.

**Review by michaelkirk — APPROVED** [review]
LGTM

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [reply src/Messages/OWSMessageSender.m:494]: That's interesting. I'm inclined to leave the new state tracking which recipients have received the message.  I'm hoping it improves the experience when sending messages to very large groups (100 or 1,000 members) from bad connections, where the odds of sending failures is high - by tracking "sent recipients", we can "resume" the send from where the error took place.  WDYT?

Comment (michaelkirk) [reply src/Messages/OWSMessageSender.m:494]: Sounds good.

File: src/Messages/Interactions/TSOutgoingMessage.m
Comment (charlesmchen) [reply src/Messages/Interactions/TSOutgoingMessage.m]: Done.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [reply src/Messages/OWSMessageSender.m]: > The downside is that we'll confuse the situation where multiple recipients do get it but, e.g. the one unregistered group member, does not get it.

You raise an excellent point.  I thought about this overnight and came up with some further changes: https://github.com/WhisperSystems/SignalServiceKit/pull/173/commits/a7df1c18d0eb515bfe34ecb3fc1b05f313811573

* I've moved the "is retryable" flag from a separate boolean in the failure handler to being a property of the error.  This lets us propagate the flag from the inner loop to the outer loop.
* I've added a "error should be ignored for group sends" flag.  So far I'm only using it for the case you mentioned (sending group messages shouldn't appear to fail if one group member unregisters), but I wonder if we should use it for any other kind of error as well.
* I've also made a number of small bug fixes.  I'll call them out with inline comments.

File: src/Messages/OWSMessageSender.h
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.h:23]: This flag has moved to the `NSError` category below.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:567]: Just some tidying up here.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:608]: Added a comment.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m]: Instead of marking all group failures as retry-able, I think we should propagate the original isRetryable flag...

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:739]: ...But there are some errors that we want to ignore when send messages to groups.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [inline src/Messages/OWSMessageSender.m:927]: A key change.

File: src/Messages/OWSMessageSender.m
Comment (charlesmchen) [reply src/Messages/OWSMessageSender.m]: @michaelkirk Hmm, looking over this last commit, I'm having second thoughts about whether moving the `isRetryable` flag from being a block parameter to a NSError property was a good idea.  Originally I thought it'd be a good idea for two reasons:

* I thought I might split it out into separate `isRetryableForContacts` and `isRetryableForGroups` flags, but my intuition was wrong - I don't think there aren't any scenarios where we want to retry for contacts but not for groups or vice versa.  
* It'd be consistent with the `shouldBeIgnoredForGroups` property.  But since we're using that in so few place,s who cares?

Anyway, I might revert most of this commit.

Comment (charlesmchen) [reply src/Messages/OWSMessageSender.m]: Okay, I've thought it over and there is a solid reason to hang the `isRetryable` state on the error - we want to look at it when resolving possibly multiple errors for a group send into a single error.  If any of them are retryable, we want to retry.  I'm going to add another commit that does this.

File: src/Messages/Interactions/TSOutgoingMessage.m
Comment (michaelkirk) [inline src/Messages/Interactions/TSOutgoingMessage.m]: Is this like a deprecation warning? Dare you just delete the method?

or at least add a comment.

File: src/Messages/OWSMessageSender.m
Comment (michaelkirk) [inline src/Messages/OWSMessageSender.m]: nit: copied typo, but :+1: on the tidying.

File: src/Messages/OWSMessageSender.m
Comment (michaelkirk) [inline src/Messages/OWSMessageSender.m]: nit: why remove this log? It's unlikely to be filling the logs up, and if it were I'd almost certainly want to know.

File: src/Messages/OWSMessageSender.m
Comment (michaelkirk) [inline src/Messages/OWSMessageSender.m]: I don't think it's a big deal, but when you're in a group with *only* unregistered users, I expect we'll see the message successfully marked as "sent" even though we sent it to noone.

I don't think this is that big of a deal, since the sender will still see all the "unregistered user" warnings, they will likely not be that "surprised" that their message wasn't sent, but it's a little weird.

Would it be straight forward to do something like:
```
 if (message.sentRecipients.count == 0) {
    // do we need to mark it as non-retryable?
    NSError *error = OWSErrorMakeNoValidRecipientsError()
    failureHandler(errror)
}
```

edit: To restate, if this isn't as straight forward to address as I think, it might not be worth addressing.

File: src/Messages/OWSMessageSender.m
Comment (michaelkirk) [inline src/Messages/OWSMessageSender.m:1008]: Not your bug, but shouldn't this actually just be "sent"?

We have no way of knowing if it was delivered to Bob yet, right?

**Review by michaelkirk — APPROVED** [review]
Just minor feedback. LGTM

**Comment by charlesmchen** [comment]
> I don't think it's a big deal, but when you're in a group with only unregistered users, I expect we'll see the message successfully marked as "sent" even though we sent it to no-one.

It's a strange edge case; I'm not sure what the desired behavior should be, so I'll defer to you.

> Would it be straight forward to do something like:

Done.

**Comment by charlesmchen** [comment]
> nit: copied typo, but 👍 on the tidying.

Done.

**Comment by charlesmchen** [comment]
> nit: why remove this log? It's unlikely to be filling the logs up, and if it were I'd almost certainly want to know.

Done.

**Comment by charlesmchen** [comment]
> Is this like a deprecation warning? Dare you just delete the method?

I was going to try to deprecate some of these initializers, but it's just not worth it. Reverted.

**Comment by charlesmchen** [comment]
> Not your bug, but shouldn't this actually just be "sent"?

I'm going to add a card so that we discuss this.

**Pull request merged**
